### PR TITLE
Decomp for aten.im2col

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84303

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -600,6 +600,63 @@ def _log_softmax_backward_data(
...
+@register_decomposition(aten.im2col)
+def im2col(
+    input: Tensor,
+    kernel_size: List[int],
+    dilation: List[int],
+    padding: List[int],
+    stride: List[int],
+) -> Tensor:
+
+    batch_dim = input.size(0)
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:612]: should you do some error checking here, like check that input dimension is 4d, kernel_size etc contain 2 elements?

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -600,6 +600,63 @@ def _log_softmax_backward_data(
...
+
+    blocks_row_indices, row_indices_len = _get_im2col_indices_along_dim(
+        input_h, kernel_h, dilation_h, padding_h, stride_h
+    )
+    blocks_col_indices, col_indices_len = _get_im2col_indices_along_dim(
+        input_w, kernel_w, dilation_w, padding_w, stride_w
+    )
+
+    padded_input = F.pad(input, (padding_w, padding_w, padding_h, padding_h))
+    output = padded_input[:, :, blocks_row_indices, :]
```
Comment (ngimel) [inline torch/_decomp/decompositions.py:652]: you should be able to combine these into a single indexing call by unsqueezeing your blocks_row_indices, see upsampling_bilinear decomp

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (SherlockNoMad) [reply torch/_decomp/decompositions.py:617]: Done.

Comment (ngimel) [reply torch/_decomp/decompositions.py:617]: Final nit, use _prims_common.check for error checking https://github.com/pytorch/pytorch/blob/8cd296f6804727899b39198d1641055b64f99056/torch/_prims_common/__init__.py#L1408

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**